Code/Framework/AzToolsFramework/AzToolsFramework/ActionManager/ToolBar/EditorToolBar.cpp

```diff
@@ -13,14 +13,110 @@
+
 #include <AzCore/Serialization/SerializeContext.h>
 
 #include <QMenu>
 #include <QToolBar>
 #include <QToolButton>
+#include <QWidgetAction>
 
 namespace AzToolsFramework
 {
+    ToolBarExpanderWatcher::ToolBarExpanderWatcher(QObject* parent)
+        : QObject(parent)
+    {
+    }
+
+    void ToolBarExpanderWatcher::Initialize()
+    {
+        s_actionManagerInterface = AZ::Interface<ActionManagerInterface>::Get();
+        AZ_Assert(s_actionManagerInterface, "EditorToolBar - Could not retrieve instance of ActionManagerInterface");
+
+        s_actionManagerInternalInterface = AZ::Interface<ActionManagerInternalInterface>::Get();
+        AZ_Assert(s_actionManagerInternalInterface, "EditorToolBar - Could not retrieve instance of ActionManagerInternalInterface");
+
+        s_menuManagerInterface = AZ::Interface<MenuManagerInterface>::Get();
+        AZ_Assert(s_menuManagerInterface, "EditorToolBar - Could not retrieve instance of MenuManagerInterface");
+
+        s_menuManagerInternalInterface = AZ::Interface<MenuManagerInternalInterface>::Get();
+        AZ_Assert(s_menuManagerInternalInterface, "EditorToolBar - Could not retrieve instance of MenuManagerInternalInterface");
+    }
+
+    bool ToolBarExpanderWatcher::eventFilter(QObject* obj, QEvent* event)
+    {
+        switch (event->type())
+        {
+        case QEvent::MouseButtonPress:
+        case QEvent::MouseButtonRelease:
+        case QEvent::MouseButtonDblClick:
+            {
+                if (qobject_cast<QToolButton*>(obj))
+                {
+                    auto mouseEvent = static_cast<QMouseEvent*>(event);
+                    auto expander = qobject_cast<QToolButton*>(obj);
+                    expander->setPopupMode(QToolButton::InstantPopup);
+
+                    auto toolbar = qobject_cast<QToolBar*>(expander->parentWidget());
+                    auto menu = new QMenu(expander);
+
+                    // Create a parent widget to more easily delete widget action widgets.
+                    auto parentWidget = new QWidget();
+
+                    for (auto action : toolbar->actions())
+                    {
+                        if (!toolbar->widgetForAction(action)->isVisible())
+                        {
+                            if (auto widgetAction = qobject_cast<QWidgetAction*>(action))
+                            {
+                                if (auto toolButton = qobject_cast<QToolButton*>(widgetAction->defaultWidget());
+                                    toolButton && toolButton->menu())
+                                {
+                                    menu->addMenu(s_menuManagerInternalInterface->GetMenu(action->objectName().toStdString().c_str()));
+                                }
+                                else
+                                {
+                                    if (QWidget* widget = s_actionManagerInternalInterface->GenerateWidgetFromWidgetAction(
```

Comment:
suggestion: Maybe could also move some of this logic to a separate function (again to help with the nesting)

Reply:
I tried to adjust the nesting a little bit, but I wanted to avoid additional functions if possible. It should be more readable now tho.

Reply:
Okay understood, thanks for having another look at it 👍